public/igameevents.h

```diff
@@ -1,187 +1,218 @@
+	}
+
+All events must have unique names (case sensitive) and may have a list
+of data fields. each data field must specify a data type, so the engine
+knows how to serialize/unserialize that event for network transmission.
+Valid data types are string, float, long, short, byte & bool. If a 
+data field should not be broadcasted to clients, use the type "local".
+*/
+
+struct GameEventKeySymbol_t
+{
+	inline GameEventKeySymbol_t(const char* keyName): m_nHashCode(0), m_pszKeyName(NULL)
+	{		
+		if (!keyName || !keyName[0])
+			return;
+
+		CUtlString buf( keyName );
+
+		buf.ToLowerFast();
+
+		m_nHashCode = MurmurHash2(buf.Get(), strlen(keyName), 0x31415926);
+		m_pszKeyName = keyName;
+
+#if 0
+		if (g_bUpdateStringTokenDatabase)
+		{
+			RegisterStringToken(keySymbol, keyName, 0, true);
+		}
+#endif
+	}
+
+private:
+	unsigned int m_nHashCode;
+	const char* m_pszKeyName;
+};
+
+
+#define MAX_EVENT_NAME_LENGTH	32		// max game event name length
+#define MAX_EVENT_BITS			9		// max bits needed for an event index
+#define MAX_EVENT_NUMBER		(1<<MAX_EVENT_BITS)		// max number of events allowed
+#define MAX_EVENT_BYTES			1024	// max size in bytes for a serialized event
+
+class KeyValues;
+class CGameEvent;
+
+abstract_class IToolGameEventAPI
+{
+	virtual void unk001( void * ) = 0;
+};
+
+abstract_class IGameEvent
+{
+public:
+	virtual ~IGameEvent() {};
+	virtual const char* GetName() const = 0;	// get event name
+	virtual int GetID() const = 0;
+
+	virtual bool IsReliable() const = 0; // if event handled reliable
+	virtual bool IsLocal() const = 0; // if event is never networked
+	virtual bool IsEmpty(const GameEventKeySymbol_t& keySymbol) = 0; // check if data field exists
+
+	// Data access
+	virtual bool  GetBool(const GameEventKeySymbol_t& keySymbol, bool defaultValue = false) = 0;
+	virtual int GetInt(const GameEventKeySymbol_t& keySymbol, int defaultValue = 0) = 0;
+	virtual uint64 GetUint64(const GameEventKeySymbol_t& keySymbol, uint64 defaultValue = 0) = 0;
+	virtual float GetFloat(const GameEventKeySymbol_t& keySymbol, float defaultValue = 0.0f) = 0;
+	virtual const char* GetString(const GameEventKeySymbol_t& keySymbol, const char* defaultValue = "") = 0;
+	virtual void* GetPtr(const GameEventKeySymbol_t& keySymbol, void* defaultValue = NULL) = 0;
+
+	/* These function prototypes and names are very speculative and might be incorrect */
+	virtual CEntityHandle GetEHandle(const GameEventKeySymbol_t& keySymbol, CEntityHandle defaultValue) = 0;
+	virtual CEntityHandle GetStrictEHandle(const GameEventKeySymbol_t& keySymbol, CEntityHandle defaultValue) = 0;
+	virtual CEntityHandle GetEHandle2(const GameEventKeySymbol_t& keySymbol, CEntityHandle defaultValue) = 0;
+
+	virtual void GetPlayerSlot(CPlayerSlot &slot, const GameEventKeySymbol_t& keySymbol) = 0;
+
+	virtual void* GetPlayerController(const GameEventKeySymbol_t& keySymbol) = 0;
+	virtual void* GetPlayerPawn(const GameEventKeySymbol_t& keySymbol) = 0;
```

Comment:
Might as well make these return `CBasePlayerController*`

Reply:
Why? We won't be having these implemented at all, just return ``IHandleEntity *`` there and let others make a use of this?

Reply:
I see, in that case it makes more sense to return what you said.